firmware/application/apps/ui_flash_utility.cpp

```diff
@@ -65,7 +76,43 @@ void FlashUtilityView::firmware_selected(std::filesystem::path::string_type path
         });
 }
 
+bool FlashUtilityView::endsWith(const std::u16string& str, const std::u16string& suffix) {
+    if (str.length() >= suffix.length()) {
+        std::u16string endOfString = str.substr(str.length() - suffix.length());
+        return endOfString == suffix;
+    } else {
+        return false;
+    }
+}
+
+std::filesystem::path FlashUtilityView::extract_tar(std::filesystem::path::string_type path) {
+    //
+    ui::Painter painter;
+    painter.fill_rectangle(
+        {0, 0, portapack::display.width(), portapack::display.height()},
+        ui::Color::black());
+    painter.draw_string({12, 24}, this->nav_.style(), "Unpacking TAR file...");
+
+    auto res = UnTar::untar(path, [this](const std::string fileName) {
+        ui::Painter painter;
+        painter.fill_rectangle({0, 50, portapack::display.width(), 90}, ui::Color::black());
+        painter.draw_string({0, 60}, this->nav_.style(), fileName);
+    });
+    if (res.string().empty()) {
+        ui::Painter painter;
+        painter.fill_rectangle({0, 50, portapack::display.width(), 90}, ui::Color::black());
+        painter.draw_string({0, 60}, this->nav_.style(), "BAD TAR FILE");
+        chThdSleepMilliseconds(5000);
+    }
+    return res;
+}
+
 void FlashUtilityView::flash_firmware(std::filesystem::path::string_type path) {
+    if (endsWith(path, u".ppfw.tar")) {
+        // extract, then update
+        path = extract_tar(u'/' + path).native();
+        if (path.empty()) return;
```

Comment:
In this case it'll go back to the flash confirmation screen without any error message, like when there is no .bin in the tar

Reply:
The message is handled in extract_tar (ui_flash_utility.cpp line 101) that shows an error message for 5 sec, then returns to the file selector screen. Tried to push another ModalMessageView, but it didn't showed up, I think because it was called when another one is already shown. This is why the 5 sec delayed message.

Reply:
Yeah now it works, before it didn't show 👍

A button to ack would be useful instead of the delay but I can live with this 😉